[Docs] Polish SkyServe update docs a bit. (#3042)
skypilot-org/skypilot: docs/source/serving/update.rst

.. _serve-update:

Updating a Service
==================

SkyServe supports *updating* a deployed service, which can be used to change:

* Replica code (e.g., ``run``/``setup``; useful for debugging)

> * Replica resource spec (e.g., number of replicas; accelerator/instance used)
> * Service spec (e.g., autoscaling spec in ``replica_policy``)

Reviewer: ```suggestion
* Replica resource spec (e.g., accelerator/instance used)
* Service spec (e.g., number of replicas; autoscaling spec in ``replica_policy``)
```
#replicas is a service spec?

During an update, the service endpoint will remain the same and the service will
remain accessible with no downtime.

Use ``sky serve update`` to update an existing service:

.. code-block:: console

    $ sky serve update service-name new_service.yaml

SkyServe will launch new replicas described by ``new_service.yaml`` with the following behavior:

> * An update is initiated, and traffic will still be redirected to existing (old) replicas.

Reviewer 2: ```suggestion
* An update is initiated, and traffic will be redirected to existing (old) replicas as new replicas are provisioning
```

Author: Do you mean "since new replicas are still provisioning we have to send traffic to old", or "..while new replicas are being provisioning"? If the latter, it should be covered by the following 2 bullets.

Reviewer 2: The former. I think the current one looks fine. Maybe remove  'still'

Author: Reworded a bit to "traffic will continue to be redirected to existing (old) replicas." which seems better.

* New replicas (with new settings) are brought up in the background.

> * Once ``min_replicas`` new replicas are ready, traffic is *switched*: traffic will be sent to the new

Author: Thanks. Guess I didn’t understand the full picture. Is this correct: 

Let’s say enough new replicas just become ready; then new requests come in which will be redirected to new replicas. However, it’s possible that old replicas are still busy servicing some old requests. 

Does SkyServe wait for old requests to be completely finished (old replicas drained) before scaling down old replicas?

> * Once ``min_replicas`` new replicas are ready, traffic is *switched*: traffic will be sent to the new

Reviewer: For now the requests processing on the old replicas will be interrupted; we introduced node draining in #2793 in which it will wait 120 seconds before terminate the node.

> * Once ``min_replicas`` new replicas are ready, traffic is *switched*: traffic will be sent to the new

Author: Reworded.

  replicas, and old replicas will stop receiving traffic and will be scaled down.
* Replicas are versioned (starting from 1). During an update, traffic is
  entirely serviced by either old-versioned or new-versioned replicas.

For example, suppose we have a running service hosting a Llama2 model with the following resource configuration:

.. code-block:: yaml

    resources:
      memory: 32+
      accelerators: T4

> You can update it to use a new resource configuration for each replica, such as:

Reviewer 2: This reads a bit like we can potentially have a different resource configuration for each replica.

Author: Reworded.

    http-server  1        1m 41s  READY   2/2       44.206.240.249:30002

    Service Replicas
    SERVICE_NAME  ID  VERSION  IP              LAUNCHED    RESOURCES       STATUS  REGION
    http-server   1   1        54.173.203.169  2 mins ago  1x AWS(vCPU=2)  READY   us-east-1
    http-server   2   1        52.87.241.103   2 mins ago  1x AWS(vCPU=2)  READY   us-east-1

Service ``http-server`` has an initial version of 1.

Suppose we want to update the service to use 4 vCPUs instead of 2. We can update
the task yaml ``examples/serve/http_server/task.yaml``, by changing the ``cpu``
field:

.. code-block:: yaml
    :emphasize-lines: 10

    # examples/serve/http_server/task.yaml
    service:
      readiness_probe:
        path: /health
        initial_delay_seconds: 20
      replicas: 2

    resources:
      ports: 8081
      cpus: 4+

    workdir: examples/serve/http_server

    run: python3 server.py

We can then use :code:`sky serve update` to update the service:

.. code-block:: console

    $ sky serve update http-server examples/serve/http_server/task.yaml

SkyServe will first launch two new replicas with 4 vCPUs. When the number of new

> replicas reaches the ``min_replicas`` (defaults to ``service.replicas`` when unspecified; i.e., 2) required for the service,

Reviewer 2: ```suggestion
replicas reaches the ``min_replicas`` (set to ``service.replicas`` when unspecified; i.e., 2) required for the service,
```

SkyServe will scale down old replicas to save cost. The service's version is
updated from 1 to 2. Replicas 3 and 4 are the new replicas with 4
vCPUs. Replicas 1 and 2 are the old replicas with 2 vCPUs. When the
new replicas are still provisioning, SkyServe will only send traffic to the old

> replicas.

Reviewer: Change to sth like "When the new replicas are not reached min_replicas"? There is some case when one replica is still provisioning while other is ready

Author: Reworded and reordered some sentences.
